### PR TITLE
core: Add float-semantics to support low-precision

### PR DESCRIPTION
**Pull request by n-io**
Extends support for reduced-precision floating point values by adding support for reduced-precision literals, and a central mechanism to implement a variety of different flavours. The types itself were added in #6250. This PR adds support for literals, by introducing the `FloatSemantics` class based on MLIR's `fltSemantics` concept (alongside `fltNonfiniteBehavior` and `fltNanEncoding`).

Stack:
* #6250 
* #6256 
* #6263

### CODE REVIEW

File: tests/dialects/test_builtin.py [hunk 1]
```diff
@@ -80,6 +82,8 @@
 from xdsl.printer import Printer
 from xdsl.utils.exceptions import VerifyException
 
+# pyright: reportPrivateUsage=false
```
Comment (superlopuh) [inline tests/dialects/test_builtin.py]: Why not expose this API?

**Review by superlopuh — COMMENTED** [review]
This is quite a lot of code to review at a time, let's split this up? I feel like the first thing to add is the overall structure, and none of the methods on `_ReducedPrecisionFloatType` (which I would make public). Then the next PRs can one-by-one replace the existing methods on the subclasses, like `bitwidth` could be one PR, and then `format` could be another PR. Does this sound like a reasonable plan?

File: tests/dialects/test_builtin.py (hunk 1 above)
Comment (n-io) [reply tests/dialects/test_builtin.py]: Changed as suggested.

**Comment by n-io** [comment]
Definitely happy to split off the `FloatSemantics` + enums, a basically empty `ReducedPrecisionFloatType` (now public), and the rewiring into the classes. It helps reduce the diff, but it's mostly optical and its harder to split up the logic. Separate PR raised, this is WIP PR for now.

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -1213,8 +1211,196 @@ def bitwidth(self) -> int:
...
+        """Round `significand >> shift` to nearest, ties to even, with exact integer math."""
+        if shift <= 0:
+            return significand << (-shift)
+        quotient = significand >> shift
+        dropped = significand & ((1 << shift) - 1)
+        halfway = 1 << (shift - 1)
+        if dropped > halfway or (dropped == halfway and quotient & 1):
+            quotient += 1
+        return quotient
+
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:1265]: Let's add these in a separate PR, also adding them to FloatSemantics with dedicated pytests?

File: xdsl/dialects/builtin.py [hunk 3]
```diff
@@ -1240,8 +1238,177 @@ def bitwidth(self) -> int:
...
+            return -math.inf if negative else math.inf
+        if semantics.nonfinite is FloatNonfiniteBehavior.FINITE_ONLY:
+            return None
+        if semantics.nan_encoding is FloatNanEncoding.ALL_ONES:
+            if (
+                exponent == semantics.max_exponent
+                and mantissa == semantics.max_mantissa
+            ):
+                return math.nan
+            return None
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: switch vibes

File: xdsl/dialects/builtin.py [hunk 4]
```diff
@@ -1240,8 +1238,177 @@ def bitwidth(self) -> int:
...
+                semantics.max_exponent << semantics.mantissa_bits
+            )
+        if semantics.nonfinite is FloatNonfiniteBehavior.NAN_ONLY:
+            return self._encode_nan(sign)
+        return (
+            (sign << semantics.sign_shift)
+            | (semantics.max_exponent << semantics.mantissa_bits)
+            | semantics.max_mantissa
+        )
+
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: also could be a switch

File: xdsl/dialects/builtin.py [hunk 5]
```diff
@@ -1240,8 +1238,177 @@ def bitwidth(self) -> int:
...
+            if quotient < implicit_bit:
+                return 0, quotient
+            return 1, 0  # rounded up into the smallest normal
+        quotient = semantics.round_half_even(
+            significand, (exponent - mantissa_bits) - power
+        )
+        if quotient == implicit_bit << 1:
+            exponent += 1
+            quotient = implicit_bit
+        exponent_field = exponent + bias
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: why `exponent_field` and not `exponent`?

File: xdsl/dialects/builtin.py [hunk 6]
```diff
@@ -1240,8 +1238,177 @@ def bitwidth(self) -> int:
...
+        if semantics.nonfinite is FloatNonfiniteBehavior.IEEE:
+            return exponent_field >= semantics.max_exponent
+        if semantics.nan_encoding is FloatNanEncoding.ALL_ONES:
+            return exponent_field > semantics.max_exponent or (
+                exponent_field == semantics.max_exponent
+                and mantissa == semantics.max_mantissa
+            )
+        return exponent_field > semantics.max_exponent
+
+    def _round_to_fields(self, magnitude: float) -> tuple[int, int] | None:
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: this is a bit of a cryptic name, how about `_rounded_exponent_mantissa`

File: xdsl/dialects/builtin.py [hunk 7]
```diff
@@ -1240,8 +1238,177 @@ def bitwidth(self) -> int:
         return self.SEMANTICS.bitwidth
 
     @property
-    def format(self) -> str:
-        raise NotImplementedError()
+    def compile_time_size(self) -> int:
+        return self.size
+
+    @staticmethod
+    def f64_significand(magnitude: float) -> tuple[int, int]:
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]:
```suggestion
    def decode_significand_power(magnitude: float) -> tuple[int, int]:
```

**Review by superlopuh — COMMENTED** [review]
Just thinking about this a bit more, it feels like a kind of `FloatStructure` would help here quite a lot, some kind of representation of a floating-point number with a given semantics. Quite a lot of the logic remaining in this PR could then be viewed as

1. unpack an f64 into its representation
2. convert this representation into a lower-precision one (potentially raising an error if it's not exactly representable)

Does this make sense? It would also pave the way towards larger floats, as those would already be representable by `FloatStructure`.

File: xdsl/dialects/builtin.py (hunk 3 above)
Comment (n-io) [reply xdsl/dialects/builtin.py]: Done as suggested.

File: xdsl/dialects/builtin.py (hunk 4 above)
Comment (n-io) [reply xdsl/dialects/builtin.py]: Done as suggested.

File: xdsl/dialects/builtin.py (hunk 5 above)
Comment (n-io) [reply xdsl/dialects/builtin.py]: `biased_exponent` is probably most precise, renamed throughout.

**Comment by n-io** [comment]
> Just thinking about this a bit more, it feels like a kind of `FloatStructure` would help here quite a lot, some kind of representation of a floating-point number with a given semantics. Quite a lot of the logic remaining in this PR could then be viewed as
> 
> 1. unpack an f64 into its representation
> 2. convert this representation into a lower-precision one (potentially raising an error if it's not exactly representable)
> 
> Does this make sense? It would also pave the way towards larger floats, as those would already be representable by `FloatStructure`.

This raises a good point in terms of matching APFloat, but it would go beyond the targeted scope, and the heavy-lifting logic is approximately the same. I'd prefer to frame this as a follow-up rather than a prequel.

Note, that APFloat rounds instead of raises.

File: xdsl/dialects/builtin.py [hunk 8]
```diff
@@ -1240,8 +1238,191 @@ def bitwidth(self) -> int:
...
+
+    def pack_into(self, buffer: WriteableBuffer, offset: int, value: float) -> None:
+        size = self.size
+        memoryview(buffer)[offset : offset + size] = self.encode_bits(value).to_bytes(
+            size, "little"
+        )
+
+    def pack(self, values: Sequence[float]) -> bytes:
+        size = self.size
+        return b"".join(self.encode_bits(v).to_bytes(size, "little") for v in values)
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: for all of these, I would probably leverage the struct package to encode/decode ints, instead of this custom logic to do the same thing, unless I'm missing something.

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: Also, it would be worth making sure that we roundtrip with MLIR in terms of roundtripping dense arrays of these floats.

File: tests/filecheck/mlir-conversion/with-mlir/dialects/builtin/builtin_reduced_fp_types.mlir [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+// CHECK: module {
+
+arith.constant 1.5 : tf32
+// CHECK-NEXT: {{%.*}} = arith.constant 1.500000e+00 : tf32
+
+arith.constant 1.5 : f8E5M2
+// CHECK-NEXT: {{%.*}} = arith.constant 1.500000e+00 : f8E5M2
+
+// 0.1 is not representable in f8E5M2; it narrows to 0.09375.
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/builtin/builtin_reduced_fp_types.mlir:17]: why have such a test only for some of the types?

File: xdsl/dialects/builtin.py (hunk 8 above)
Comment (n-io) [reply xdsl/dialects/builtin.py]: The struct package does not work well for instance with the 3-byte `tf32` type, as the integer codes are 1/2/4/8 bytes only.

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: I had a feeling there might be something tricky like this going on, this feels worth testing specifically, and adding more documentation. At the same time, an approach similar to what we used for bf16 might be appropriate here, where we have a 4-byte scratchpad bit of memory that we save the integer to, and then we copy just three bytes out on every iteration?

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: Also, I would express `pack` and `unpack` in terms of `pack_into` and `iter_unpack`

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: Ah sorry you already do that in `unpack`, I mean it would still be nice to use something like `return itertools.islice(self.iter_unpack(buffer), num)`

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: Actually sorry, I'm now thinking it might be worth doing the opposite, I'll play with this locally later today.

Comment (n-io) [reply xdsl/dialects/builtin.py]: I've pushed a commit, see if you like it or we can undo this again.

File: tests/dialects/test_builtin.py [hunk 10]
```diff
@@ -168,23 +170,9 @@ def test_FloatType_formats():
...
-        f6E2M3FN,
-        f6E3M2FN,
-        f4E2M1FN,
-    )
-    for float_type in reduced_precision_floats:
-        with pytest.raises(NotImplementedError):
-            float_type.format
+    # Reduced-precision floats pack via the reduced-float codec, not a struct
+    # format string, so (like bf16) they have no `format` attribute.
+    assert not hasattr(tf32, "format")
```
Comment (superlopuh) [inline tests/dialects/test_builtin.py]: This feels off to me, what are we really testing here?

Comment (superlopuh) [reply tests/dialects/test_builtin.py]: My gut tells me to just delete this line, I'm not sure what benefit there is to it.

File: xdsl/dialects/builtin.py [hunk 11]
```diff
@@ -1240,8 +1239,195 @@ def bitwidth(self) -> int:
...
+            | (biased_exponent << semantics.mantissa_bits)
+            | mantissa
+        )
+
+    def iter_unpack(self, buffer: ReadableBuffer, /) -> Iterator[float]:
+        # `int` packs the odd byte widths `struct` can't, e.g. tf32's 19 bits -> 3 bytes.
+        size = self.size
+        mv = memoryview(buffer)
+        for i in range(0, len(mv), size):
+            yield self.decode_bits(int.from_bytes(bytes(mv[i : i + size]), "little"))
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: Does this work?
```suggestion
            yield self.decode_bits(int.from_bytes(mv[i : i + size]), "little")
```

**Review by superlopuh — APPROVED** [review]
TYVM

**Comment by superlopuh** [comment]
If you have the motivation it feels to me like it would be great to support the larger floats also

File: tests/dialects/test_builtin.py (hunk 10 above)
Comment (n-io) [reply tests/dialects/test_builtin.py]: It's mirroring the same (pre-existing) check for bf16 just above. Either way is fine.

**Review by Samielakkad — COMMENTED** [review]
Nice addition overall. One edge case I’d want covered before this lands: `iter_unpack` currently walks the buffer in `size` chunks and decodes the last slice even if it is shorter than `size`.

For the reduced-float types with odd byte widths, that means a truncated buffer can silently decode as if the missing high bytes were zero. That is a little different from `struct.iter_unpack`, which requires an exact multiple of the element size.

It may be worth either checking `len(buffer) % self.size == 0` in `iter_unpack` / `unpack`, or adding an explicit test that this silent partial decode is the intended behavior. I would lean toward rejecting the partial chunk so malformed dense data does not quietly round-trip to a different value.

File: xdsl/dialects/builtin.py (hunk 11 above)
Comment (n-io) [reply xdsl/dialects/builtin.py]: Indeed 😃

File: tests/dialects/test_builtin.py (hunk 10 above)
Comment (n-io) [reply tests/dialects/test_builtin.py]: Moved these next to each other, removed comment. Share the general gut feeling though.

**Comment by n-io** [comment]
Thanks @Samielakkad, I think `bf16` rejects too so it makes sense to replicate that.

File: xdsl/dialects/builtin.py [hunk 12]
```diff
@@ -1240,8 +1239,202 @@ def bitwidth(self) -> int:
...
+
+    def iter_unpack(self, buffer: ReadableBuffer, /) -> Iterator[float]:
+        # `int` handles the odd byte widths `struct` can't, e.g. tf32's 19 bits -> 3 bytes.
+        size = self.size
+        mv = memoryview(buffer)
+        if len(mv) % size:
+            raise ValueError(
+                f"buffer of {len(mv)} bytes is not a multiple of the {size}-byte "
+                f"element size of {self.name}"
+            )
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:1462]: No, this is also not the right check, as it should be OK to unpack lazily just the prefix. The error should be raised only if the last element cannot be read in full

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/dialects/builtin.py (hunk 12 above)
Comment (superlopuh) [reply xdsl/dialects/builtin.py:1462]: I'm also not sure about the comment, the old code was reading potentially past the end of the buffer, not silently truncating? I would have expected an error to be raised there from an out of bounds access for the last element.

Comment (superlopuh) [reply xdsl/dialects/builtin.py:1462]: Reading the struct documentation, it says that the bytes buffer should be a multiple of the struct size but not what error is raised if it isn't...

Comment (n-io) [reply xdsl/dialects/builtin.py:1462]: I think the OOB error didn't happen because of the slicing.

File: xdsl/dialects/builtin.py [hunk 13]
```diff
@@ -1240,8 +1239,201 @@ def bitwidth(self) -> int:
...
+        size = self.size
+        mv = memoryview(buffer)
+        for i in range(0, len(mv), size):
+            chunk = mv[i : i + size]
+            if len(chunk) < size:
+                raise ValueError(
+                    f"buffer ends with a partial {self.name} value: "
+                    f"{len(chunk)} of {size} bytes"
+                )
+            yield self.decode_bits(int.from_bytes(chunk, "little"))
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py]: Thanks for updating this, but I still think it would be good to be consistent across implementations. Could you please add f64 to your test for this, and check that the error type is the same for both the struct and this manual implementation, and that the errors are consistent on whether the whole buffer needs to have size multiple of this element's size or just the last element?

Comment (n-io) [reply xdsl/dialects/builtin.py]: If we're going for consistency, we should throw a `struct.error`. However, to do so I'd like to inherit from StructPackableType. Although we're not using any of its code, we do provide the same methods, so it's somewhat cleaner. I've also parameterized 
```python
@pytest.mark.parametrize("type_", [f64, bf16, tf32])
def test_float_rejects_truncated_buffer(type_: AnyFloat):
```
to demonstrate uniform rejection errors across these types.

File: tests/dialects/test_builtin.py [hunk 14]
```diff
@@ -314,6 +288,163 @@ def test_FloatAttr_normalisation(
...
+    """
+    The struct-based (f64) and manual (bf16, tf32) float codecs all reject a truncated
+    buffer with struct.error: reading one to completion never silently decodes a partial
+    trailing element as if its missing bytes were zero.
+    """
+    packed = type_.pack((1.5, 2.0))
+    assert type_.unpack(packed, 2) == (1.5, 2.0)
+    truncated = packed[:-1]  # ends part-way through the final element
+    with pytest.raises(struct.error):
+        list(type_.iter_unpack(truncated))
```
Comment (superlopuh) [inline tests/dialects/test_builtin.py]: This seems important to me, does it raise when you fetch the first element or all the elements
```suggestion
        next(type_.iter_unpack(truncated))
```

Comment (n-io) [reply tests/dialects/test_builtin.py]: Yes, with a small fix for `bf16` to match others.

File: xdsl/dialects/builtin.py (hunk 13 above)
Comment (superlopuh) [reply xdsl/dialects/builtin.py]: Let's iterate on the PR I just opened (https://github.com/xdslproject/xdsl/pull/6273) to decide what errors we want, feels like a good order of solving these issues to me.

File: tests/dialects/test_builtin.py (hunk 14 above)
Comment (superlopuh) [reply tests/dialects/test_builtin.py]: In the PR I just opened it actually raises before you fetch the first element

**Review by Samielakkad — COMMENTED** [review]
Found one edge case in `f8E8M0FNU`: negative inputs lose their sign before rounding, so `-4.0` is encoded as `+4.0` (`0x81`). LLVM rejects signed values for this format, while `ml_dtypes` maps them to `0xff`/NaN.

We should pick one of those behaviors rather than silently taking the absolute value. Tests for `-0.0` and one negative finite value would catch it.

**Review by superlopuh — APPROVED** [review]
LGTM!

**Comment by n-io** [comment]
Thanks @superlopuh, and good spot @Samielakkad! It's what one gets for sweeping the hexadecimal spectrum (:
